core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -188,11 +190,24 @@ protected void add(DataFile file) {
    * Add a delete file to the new snapshot.
    */
   protected void add(DeleteFile file) {
-    addedFilesSummary.addedFile(spec, file);
+    setWriteSpec(file);
+    addedFilesSummary.addedFile(writeSpec(), file);
     hasNewDeleteFiles = true;
     newDeleteFiles.add(file);
   }
 
+  private void setWriteSpec(ContentFile<?> file) {
+    PartitionSpec writeSpec = ops.current().spec(file.specId());
+    Preconditions.checkNotNull(file, "Invalid content file: null");
+    Preconditions.checkNotNull(writeSpec,
+        "Partition spec id should be defined in table, writing partition spec: null");
```

Comment:
Our error messages are usually of the form "Cannot <action>: <reason>". This would be "Cannot find partition spec for file: %s"